### PR TITLE
localai: init at 2.8.0

### LINKED ISSUE #269582: Package request: LocalAI
**Project description**

The free, Open Source OpenAI alternative. Self-hosted, community-driven and local-first. Drop-in replacement for OpenAI running on consumer-grade hardware. No GPU required. Runs ggml, gguf, GPTQ, onnx, TF compatible models: llama, llama2, rwkv, whisper, vicuna, koala, cerebras, falcon, dolly, starcoder, and many others

**Metadata**

* homepage URL: https://localai.io/
* source URL: https://github.com/mudler/LocalAI
* license: mit
* platforms: linux

**Priorities**

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/269582

ToDo

- [x] Add [ck3d](https://github.com/ck3d) as maintainer
- [ ] Add module
- [ ] Add tests
- [ ] Add [ck3d](https://github.com/ck3d) and others as co-authors

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/lo/localai/package.nix [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  modBuildPhase = ''
+    mkdir sources
+    make prepare-sources
+    go mod tidy -v
+  '';
+
+  proxyVendor = true;
+
+  buildPhase = ''
+    mkdir sources
```
Comment (teto) [inline pkgs/by-name/lo/localai/package.nix:192]: you should call the hooks, e.g. `runHook preBuild`

Comment (teto) [reply pkgs/by-name/lo/localai/package.nix:192]: can this be done via `makeFlags` and the inherited buildPhase ? doesn't upstream run `mkdir sources` ?

File: pkgs/by-name/lo/localai/package.nix [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+, spdlog
+, fmt
+, onnxruntime
+, cmake
+, buildGoModule
+, pkg-config
+, cudaPackages
+, makeWrapper
+, runCommand
+, buildType ? ""
```
Comment (teto) [inline pkgs/by-name/lo/localai/package.nix:27]: `buildType` deserves a comment about its possible values

Comment (ck3d) [reply pkgs/by-name/lo/localai/package.nix:27]: Possibles values are defined here:
https://github.com/mudler/LocalAI/blob/master/docs/content/docs/getting-started/build.md?plain=1#L86

this derivation handles following buildTypes: "" "cublas" "openblas"
Not supported is: "hipblas" "clblas" "szcl_f16" "szcl_f32" "metal"

File: pkgs/by-name/lo/localai/package.nix [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    rev = "4b7b38bef5addbd31f453871d79647fbae6bec8a";
+    hash = "sha256-HkfvhgJ9nPIoNF1LHrEr/ICUNvGCm7lzZpqs7pQY/cY=";
+    fetchSubmodules = true;
+  };
+
+  llama_cpp' = runCommand "llama_cpp_src" { } ''
+    cp -r --no-preserve=mode,ownership ${llama_cpp} $out
+    sed -i $out/CMakeLists.txt \
+      -e 's;pkg_check_modules(DepBLAS REQUIRED openblas);pkg_check_modules(DepBLAS REQUIRED openblas64);'
+  '';
```
Comment (dotlambda) [inline pkgs/by-name/lo/localai/package.nix:64]: Can we use `pkgs.substitute`?

File: pkgs/by-name/lo/localai/package.nix [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    cp -r --no-preserve=mode,ownership ${go-tiny-dream} $out
+    sed -i $out/Makefile \
+      -e 's;lib/libncnn;lib64/libncnn;g'
+  '';
+
+  GO_TAGS = lib.optional enableTinydream "tinydream"
+    ++ lib.optional enableTts "tts"
+    ++ lib.optional enableStablediffusion "stablediffusion";
+in
+(buildGoModule.override { stdenv = gcc13Stdenv; }) rec {
```
Comment (dotlambda) [inline pkgs/by-name/lo/localai/package.nix:132]: Isn't GCC 13 the default?

File: pkgs/by-name/lo/localai/package.nix (hunk 1 above)
Comment (ck3d) [reply pkgs/by-name/lo/localai/package.nix:192]: The `buildGoModule` derivation does not call `make`, therefor we have to override `buildPhase`.

The `makeFlags` could be separated, but then the custom `buildPhase` has to respect the environment variable.

Regarding directory sources, the original `Makefile` calls `git clone`, which creates the sources directory. The patched `Makefile` calls `cp`, which does not create the leading directories.

File: pkgs/by-name/lo/localai/package.nix (hunk 4 above)
Comment (ck3d) [reply pkgs/by-name/lo/localai/package.nix:132]: Yes, gcc13Stdenv can be removed

**Comment by ck3d** [comment]
@onny : I fixed some comments in my repo. May you please update this file with the latest version?

**Comment by KiaraGrouwstra** [comment]
fyi, i'm having trouble getting this to build so far:

```
nix run nixpkgs#nixpkgs-review -- pr 288269
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/288269/head:refs/nixpkgs-review/1
$ git worktree add /home/kiara/.cache/nixpkgs-review/pr-288269-1/nixpkgs 8c826e35a91ac04504d1d061d258ad8b35050bed
Preparing worktree (detached HEAD 8c826e35a91a)
Updating files: 100% (39683/39683), done.
HEAD is now at 8c826e35a91a Merge pull request #263108 from b-rodrigues/quarto_1.4
$ git merge --no-commit --no-ff 8b0990fb586099893bbad221bd7a0d923debc52e
Automatic merge went well; stopped before committing as requested
$ nix build --nix-path nixpkgs=/home/kiara/.cache/nixpkgs-review/pr-288269-1/nixpkgs nixpkgs-overlays=/tmp/tmpxcs7_lpx --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/kiara/.cache/nixpkgs-review/pr-288269-1/build.nix
error: builder for '/nix/store/xdyng4d09z6g03mrkd0xfwy2x9564vqn-localai-2.8.0.drv' failed with exit code 1;
       last 10 log lines:
       > go: downloading github.com/valyala/tcplisten v1.0.0
       > go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.4
       > go: downloading github.com/xrash/smetrics v0.0.0-20201216005158-039620a65673
       > go: downloading github.com/cpuguy83/go-md2man/v2 v2.0.2
       > go: downloading github.com/russross/blackfriday/v2 v2.1.0
       > buildPhase completed in 9 minutes 1 seconds
       > Running phase: checkPhase
       > /nix/store/5j98bvxxb4wdmpgyn9nh13w7mp7jx2ga-stdenv-linux/setup: line 1582: getGoDirs: command not found
       > Running phase: installPhase
       > install: cannot stat 'localai': No such file or directory
       For full logs, run 'nix log /nix/store/xdyng4d09z6g03mrkd0xfwy2x9564vqn-localai-2.8.0.drv'.
error: 1 dependencies of derivation '/nix/store/35p9c9kcqhbmww3fnn6zahlycpd668l1-review-shell.drv' failed to build
```

**Comment by ck3d** [comment]
@onny : I seems you are busy at the moment, so I created another PR #291647

**Comment by onny** [comment]
Thank you @ck3d

**Pull request closed without merging**